### PR TITLE
[Merged by Bors] - update archetypes for run criterias

### LINKED ISSUE #2000: Query in a run criteria system doesn't work
## Bevy version

Bevy main 38feddb87850424df3a0b08bae8dc32c57004798

## Operating system & version

Windows 10

## What you did
```rust
struct Foo;

fn number_of_entities_system(mut commands: Commands, query: Query<&Foo>) {
    println!("num entities from dummy: {}", query.iter().count());
    commands.spawn_bundle((Foo,));
}

fn number_of_entities_critiera(query: Query<&Foo>) -> ShouldRun {
    println!("num entities from criteria: {}", query.iter().count());
    ShouldRun::Yes
}

App::build()
    .add_plugins(MinimalPlugins)
    .add_system(
        number_of_entities_system
            .system()
            .with_run_criteria(number_of_entities_critiera.system()),
    )
    .run();
```

## What you expected to happen

Both systems print the same number of entities

## What actually happened

Run criteria reports zero entities, system reports correct number.

## Additional information

Console log on my machine:

```
num entities from dummy: 1695
num entities from criteria: 0
num entities from dummy: 1696
num entities from criteria: 0
num entities from dummy: 1697
num entities from criteria: 0
num entities from dummy: 1698
```

### PR DESCRIPTION
**Pull request by mockersf**
fixes #2000 

archetypes were not updated for run criteria on a stage or on a system set

### CODE REVIEW

File: crates/bevy_ecs/src/schedule/run_criteria.rs [hunk 1]
```diff
@@ -47,13 +47,16 @@ pub enum ShouldRun {
...
+    archetype_generation: ArchetypeGeneration,
 }
 
 impl Default for BoxedRunCriteria {
     fn default() -> Self {
         Self {
             criteria_system: None,
             initialized: false,
+            // MAX ensures access information will be initialized on first run.
+            archetype_generation: ArchetypeGeneration::new(usize::MAX),
```
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/run_criteria.rs:59]: Note, this conflicts with #2172

Comment (NathanSWard) [reply crates/bevy_ecs/src/schedule/run_criteria.rs:59]: With that PR, `usize::MAX` is no longer a special case for `ArchetypeGeneration`

File: crates/bevy_ecs/src/schedule/run_criteria.rs [hunk 2]
```diff
@@ -70,6 +73,21 @@ impl BoxedRunCriteria {
...
             }
+            let archetypes = world.archetypes();
+            let old_generation = self.archetype_generation;
+            let new_generation = archetypes.generation();
+            if old_generation != new_generation {
+                let archetype_index_range = if old_generation.value() == usize::MAX {
+                    0..archetypes.len()
+                } else {
+                    old_generation.value()..archetypes.len()
+                };
```
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/run_criteria.rs:84]: Same with conflict with #2172

File: crates/bevy_ecs/src/schedule/run_criteria.rs [hunk 3]
```diff
@@ -106,6 +125,8 @@ impl RunCriteriaContainer {
             label: descriptor.label,
             before: descriptor.before,
             after: descriptor.after,
+            // MAX ensures access information will be initialized on first run.
+            archetype_generation: ArchetypeGeneration::new(usize::MAX),
```
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/run_criteria.rs:129]: ditto

File: crates/bevy_ecs/src/schedule/run_criteria.rs [hunk 4]
```diff
@@ -122,6 +143,32 @@ impl RunCriteriaContainer {
...
+        let old_generation = self.archetype_generation;
+        let new_generation = archetypes.generation();
+        if old_generation == new_generation {
+            return;
+        }
+        let archetype_index_range = if old_generation.value() == usize::MAX {
+            0..archetypes.len()
+        } else {
+            old_generation.value()..archetypes.len()
+        };
```
Comment (NathanSWard) [inline crates/bevy_ecs/src/schedule/run_criteria.rs:158]: ditto

File: crates/bevy_ecs/src/schedule/run_criteria.rs (hunk 1 above)
Comment (mockersf) [reply crates/bevy_ecs/src/schedule/run_criteria.rs:59]: yup, this PR fixes the issue as it is in current main, I'll update it if #2172 is merged first

Comment (mockersf) [reply crates/bevy_ecs/src/schedule/run_criteria.rs]: updated

**Review by alice-i-cecile — APPROVED** [review]

**Review by Ratysz — APPROVED** [review]
Oh, hm. I guess exclusive systems suffer from the same bug?

**Comment by cart** [comment]
Yup I've already fixed the exclusive system bug in the pipelined rendering branch. Planned on upstreaming but never got around to it :)

```rust
    fn run(&mut self, world: &mut World) {
        let archetypes = world.archetypes();
        let new_generation = archetypes.generation();
        let old_generation = std::mem::replace(&mut self.archetype_generation, new_generation);
        let archetype_index_range = old_generation.value()..new_generation.value();

        for archetype in archetypes.archetypes[archetype_index_range].iter() {
            self.system.new_archetype(archetype);
        }
        self.system.run((), world);
        self.system.apply_buffers(world);
    }
```

**Review by NathanSWard — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
We should include the exclusive system fix in the same PR right?

**Comment by mockersf** [comment]
> We should include the exclusive system fix in the same PR right?

added fix and test for exclusive systems

**Review by alice-i-cecile — APPROVED** [review]

**Review by Ratysz — APPROVED** [review]

**Pull request closed without merging**